File: LocalPackages/StoryAlign/Sources/StoryAlignCore/Ebook/Parser/EpubParser.swift
//
// EpubParser.swift
//
// SPDX-License-Identifier: MIT
// Copyright (c) 2025 Rich Waters
//

import Foundation
import ZIPFoundation


public struct EpubParser : AlignmentSessionProviding {
    public var session: AlignmentSession

    public init(session:AlignmentSession) {
        self.session = session
    }
    
    public func parse(url epubURL: URL ) async throws -> EpubDocument {
        try Task.checkCancellation()

        let fileManager = FileManager.default
        let tempDir = session.request.sessionDir.appendingPathComponent(UUID().uuidString)
        try fileManager.createDirectory(at: tempDir, withIntermediateDirectories: true)
        try fileManager.unzipItem(at: epubURL, to: tempDir)
        
        let containerPath = tempDir.appendingPathComponent("META-INF/container.xml")
        let containerData = try Data(contentsOf: containerPath)
        let container = try EpubContainerParser().parseContainer(from: containerData)
        
        let opfFullURL = tempDir.appendingPathComponent(container.opfPath)
        let opfData = try Data(contentsOf: opfFullURL)
        let metadata = try EpubMetaInfoParser().parseMetaInfo(from: opfData)
        let spine = try EpubSpineParser().parseSpine(from: opfData)
        let opfManifestItems = try EpubOpfManifestParser().parseManifest(from: opfData)
        
        session.progressTracker.updateProgress(for: .epub,  event:.start, total: opfManifestItems.count)

        
        var docResources:Set<URL> = []
        
        let nav:EpubNav? = try EpubNavParser().parseNav(from: opfFullURL, opfManifestItems: opfManifestItems )
        let ncx:Epub2Ncx? = try? EpubNcxParser().parseNcx( from:opfFullURL, opfManifestItems: opfManifestItems, spine: spine )


        let manifest = try opfManifestItems.compactMap { (opfManifestItem) -> EpubManifestItem? in
            try Task.checkCancellation()

            let itemName = nav?.title(for: opfManifestItem.href) ?? ncx?.title(for: opfManifestItem.href) ?? opfManifestItem.id

            guard let url = EpubOpfResolver.resolveHref(opfManifestItem.href, relativeTo: opfFullURL) else {
                throw StoryAlignError("Cannot find manifest item content" )
            }
            defer {
                session.progressTracker.updateProgress(for: .epub,increment: 1, item:itemName )
            }
            if opfManifestItem.mediaType != EpubMediaTypes.xhtmlXml {
                logger.log( .info, "Ignoring nonXhtmlItem in manifest: \(opfManifestItem.id)")
                return nil
            }

            logger.log(.info, "Parsing manifest item \(opfManifestItem.id)")
            let xmlData = try Data(contentsOf: url)
            let xmlText = String(data: xmlData, encoding: .utf8) ?? ""
            let (itemText, itemHasScript, itemResources) = try EpubXhtmlTextParser().parseText(from:xmlData )
            
            let itemResourceUrls = itemResources.map {
                url.deletingLastPathComponent().appendingPathComponent($0)
            }
            docResources.formUnion(itemResourceUrls)
            
            let granularity:Granularity = sessionConfig.granularity.useWordTokenizer ? .sentence : sessionConfig.granularity
            
            let xhtmlSentences = try EpubXhtmlTextParser.getXHtmlSentences(from: xmlText, granularity:granularity)

            let spineItemIndex = spine.item(forIdRef: opfManifestItem.id)?.index ?? -1

            let manifestItem = EpubManifestItem(
                id: opfManifestItem.id,
                href: opfManifestItem.href,
                mediaType: opfManifestItem.mediaType,
                properties: opfManifestItem.properties,
                spineItemIndex: spineItemIndex,
                text: itemText,
                xmlData: xmlData,
                hasScript: itemHasScript,
                xhtmlSentences: xhtmlSentences,
                name: itemName,
                filePath: url)

            return manifestItem
        }

        
        let guide = try Epub2GuideParser().parseGuide(from: opfData)
        
        let epub = try EpubDocument(opfPath: container.opfPath, opfXmlData: opfData, metaInfo: metadata, unzippedURL: tempDir, guide: guide, manifest: manifest, spine: spine, nav:nav, ncx:ncx, resources: Array(docResources))
        if !epub.isEpub2 && ((epub.nav?.landmarks.bodymatterHrefs.isEmpty) ?? true) {
            if sessionConfig.startManifestItemId == nil {
                logger.log(.warn, "EPUB is missing a 'bodymatter' marker (commonly recommended in EPUB3). If early alignment looks wrong, try configuring a start chapter.")
            }
        }
        
        progressTracker.updateProgress(for: .epub,  event: .end)
        
        return epub
    }
}



public extension EpubParser {
    static func chapterEntries( from epubURL:URL, logger:Logger ) throws -> [EpubChapterEntry]  {
        let archive = try Archive(url: epubURL, accessMode: .read)
        let containerData = try extractData(from: archive, path: "META-INF/container.xml")
        let container = try EpubContainerParser().parseContainer(from: containerData)
        
        let opfData = try extractData(from: archive, path: container.opfPath)
        let spine = try EpubSpineParser().parseSpine(from: opfData)
        let opfManifestItems = try EpubOpfManifestParser().parseManifest(from: opfData)
        let manifestById = EpubOpfResolver.manifestById(opfManifestItems)
        
        func chapters(using toc: some EpubNavOrNcx) -> [EpubChapterEntry] {
            spine.items.enumerated().compactMap { (index,spineItem) in
                guard let manifestItem = manifestById[spineItem.idref] else {
                    logger.log(.warn, "Spine references missing manifest id: \(spineItem.idref)")
                    return nil
                    //return EpubChapterEntry(manifestId: spineItem.idref, navLabel: spineItem.idref,
                }
                let title = toc.title(for: manifestItem.href)
                let resolvedTitle = title ?? spineItem.idref
                let role = toc.role(for: manifestItem.href) ?? (title == nil ? .unlisted : nil)
                return EpubChapterEntry(manifestId: spineItem.idref, navLabel: resolvedTitle, spineItemIndex: index, role: role)
            }
        }
        
        if let navItem = EpubOpfResolver.navManifestItem( opfManifestItems ) {
            let navPath = container.opfPath.deletingLastPathComponent().appendingPathComponent(navItem.href)
            let navData = try extractData(from: archive, path: navPath)
            let nav = try EpubNavParser().parseNav(from: navData,navManifestItem: navItem)
            return chapters(using: nav)
        }
        if let ncxItem = EpubOpfResolver.ncxManifestItem(opfManifestItems, spine: spine ) {
            let path = container.opfPath.deletingLastPathComponent().appendingPathComponent(ncxItem.href)
            let data = try extractData(from: archive, path: path)
            let ncx = try EpubNcxParser().parseNcx(from: data, ncxManifestItem: ncxItem)
            return chapters(using: ncx)
        }
        return []
    }

    static func extractData( from archive:Archive, path:String ) throws -> Data {
        guard let entry = archive[path] else {
            throw CocoaError(.fileReadNoSuchFile)
        }
        var data = Data()
        _ = try archive.extract(entry) { chunk in
            data.append(chunk)
        }

        return data
    }
}


struct EpubOpfResolver {
    static func baseURL(opfURL: URL) -> URL {
        opfURL.deletingLastPathComponent()
    }

    static func resolveHref(_ href: String, relativeTo opfURL: URL) -> URL? {
        URL(string: href, relativeTo: baseURL(opfURL: opfURL))
    }

    static func navManifestItem(_ items: [OpfManifestItem]) -> OpfManifestItem? {
        items.first { $0.properties?.contains("nav") == true }
    }
    static func ncxManifestItem(_ items: [OpfManifestItem], spine:EpubSpine ) -> OpfManifestItem? {
        if !spine.toc.isEmpty {
            if let item = items.first(where:({ $0.id == spine.toc })) {
                return item
            }
        }
        if let item = items.first(where:({ $0.mediaType == EpubMediaTypes.ncxXml })) {
            return item
        }
        return items.first { $0.properties?.contains("ncx") == true }
    }

    static func navURL(opfURL: URL, manifestItems: [OpfManifestItem]) -> URL? {
        guard let navHref = navManifestItem(manifestItems)?.href else { return nil }
        return baseURL(opfURL: opfURL).appendingPathComponent(navHref)
    }

    static func manifestById(_ items: [OpfManifestItem]) -> [String: OpfManifestItem] {
        Dictionary(uniqueKeysWithValues: items.map { ($0.id, $0) })
    }

    static func spineURLs(opfURL: URL, manifestItemsById: [String: OpfManifestItem], spine: EpubSpine) -> [URL] {
        spine.items.compactMap { itemref in
            guard let m = manifestItemsById[itemref.idref] else { return nil }
            return baseURL(opfURL: opfURL).appendingPathComponent(m.href)
        }
    }
}

File: android/engine/src/main/java/com/enve/app/storyalign/epub/EpubParser.kt
package com.enve.app.storyalign.epub

import com.enve.engine.storyalign.StoryAlignGranularity
import java.io.File
import java.io.InputStream
import java.util.zip.ZipInputStream

class EpubZip private constructor(private val entries: Map<String, ByteArray>) {
    fun bytes(path: String): ByteArray? = entries[path]
    fun entries(): Map<String, ByteArray> = entries

    companion object {
        fun from(input: InputStream): EpubZip {
            val map = LinkedHashMap<String, ByteArray>()
            ZipInputStream(input).use { zis ->
                var entry = zis.nextEntry
                while (entry != null) {
                    if (!entry.isDirectory) map[entry.name] = zis.readBytes()
                    entry = zis.nextEntry
                }
            }
            return EpubZip(map)
        }

        fun from(file: File): EpubZip = file.inputStream().buffered().use { from(it) }
    }
}

object EpubParser {

    fun parse(zip: EpubZip, granularity: StoryAlignGranularity = StoryAlignGranularity.SENTENCE): EpubDocument {
        val containerBytes = zip.bytes("META-INF/container.xml")
            ?: throw StoryAlignParseException("Missing META-INF/container.xml")
        val opfPath = parseEpubContainer(containerBytes)
        val opfBytes = zip.bytes(opfPath) ?: throw StoryAlignParseException("Missing OPF at $opfPath")
        val opfDoc = parseXmlDocument(opfBytes)
        val opfDir = opfPath.deletingLastPathComponent()

        val metaInfo = parseEpubMetaInfo(opfDoc)
        val spine = parseEpubSpine(opfDoc)
        val manifestItems = parseEpubManifest(opfDoc)

        val nav = EpubOpfResolver.navManifestItem(manifestItems)?.let { navItem ->
            zip.bytes(resolvePath(opfDir, navItem.href))?.let { parseEpubNav(it, navItem.href) }
        }
        val ncx = EpubOpfResolver.ncxManifestItem(manifestItems, spine)?.let { ncxItem ->
            zip.bytes(resolvePath(opfDir, ncxItem.href))?.let { parseEpubNcx(it, ncxItem.href, ncxItem.id) }
        }

        val parseGranularity = if (granularity.useWordTokenizer) StoryAlignGranularity.SENTENCE else granularity
        val resources = LinkedHashSet<String>()

        val manifest = manifestItems.mapNotNull { item ->
            if (item.mediaType != EpubMediaTypes.XHTML_XML) return@mapNotNull null
            val path = resolvePath(opfDir, item.href)
            val bytes = zip.bytes(path)
                ?: throw StoryAlignParseException("Cannot find manifest item content: ${item.href}")
            val xhtml = String(bytes, Charsets.UTF_8)
            val (text, hasScript, itemResources) = EpubXhtmlTextParser.parseText(xhtml)
            val itemDir = path.deletingLastPathComponent()
            itemResources.forEach { resources.add(resolvePath(itemDir, it)) }
            val name = nav?.title(item.href) ?: ncx?.title(item.href) ?: item.id
            val sentences = EpubXhtmlTextParser.getXHtmlSentences(xhtml, parseGranularity)
            val spineIndex = spine.item(item.id)?.index ?: -1
            EpubManifestItem(
                id = item.id,
                href = item.href,
                mediaType = item.mediaType,
                properties = item.properties,
                spineItemIndex = spineIndex,
                text = text,
                hasScript = hasScript,
                xhtmlSentences = sentences,
                name = name,
                xhtml = xhtml,
            )
        }

        val guide = parseEpubGuide(opfDoc)
        return EpubDocument(
            opfPath = opfPath,
            metaInfo = metaInfo,
            guide = guide,
            manifest = manifest,
            spine = spine,
            nav = nav,
            ncx = ncx,
            resources = resources.toList(),
        )
    }

    fun chapterEntries(zip: EpubZip): List<EpubChapterEntry> {
        val containerBytes = zip.bytes("META-INF/container.xml") ?: return emptyList()
        val opfPath = parseEpubContainer(containerBytes)
        val opfBytes = zip.bytes(opfPath) ?: return emptyList()
        val opfDoc = parseXmlDocument(opfBytes)
        val opfDir = opfPath.deletingLastPathComponent()
        val spine = parseEpubSpine(opfDoc)
        val manifestItems = parseEpubManifest(opfDoc)
        val manifestById = manifestItems.associateBy { it.id }

        fun chapters(toc: EpubNavOrNcx): List<EpubChapterEntry> =
            spine.items.mapIndexedNotNull { index, spineItem ->
                val m = manifestById[spineItem.idref] ?: return@mapIndexedNotNull null
                val title = toc.title(m.href)
                val role = toc.role(m.href) ?: if (title == null) EpubChapterRole.UNLISTED else null
                EpubChapterEntry(
                    manifestId = spineItem.idref,
                    navLabel = title ?: spineItem.idref,
                    spineItemIndex = index,
                    role = role,
                )
            }

        EpubOpfResolver.navManifestItem(manifestItems)?.let { navItem ->
            zip.bytes(resolvePath(opfDir, navItem.href))?.let { return chapters(parseEpubNav(it, navItem.href)) }
        }
        EpubOpfResolver.ncxManifestItem(manifestItems, spine)?.let { ncxItem ->
            zip.bytes(resolvePath(opfDir, ncxItem.href))?.let { return chapters(parseEpubNcx(it, ncxItem.href, ncxItem.id)) }
        }
        return emptyList()
    }
}
